### PR TITLE
Dynamic commands and improved registration

### PR DESCRIPTION
**Pull request by PizzaCrust**
This commit adds registration handlers for implementations to handle command registration at any time, than just startup. This commit also adds dynamic commands called ICommand. It also fixes calling the same command name multiple times.

### CODE REVIEW

**Comment by phase** [comment]
What are the use of dynamic commands? I think having immutable ones would be safer.

**Comment by PizzaCrust** [comment]
Dynamic commands allows for plugins like Skript, to register commands dynamically after they parse the scripting language.

**Comment by minecrafter** [comment]
I suggest that interfaces not be used as they are not safe. Please use an immutable object instead.

**Comment by PizzaCrust** [comment]
Hmm. Probably like this:

``` java
Fountain.getServer().getCommandManager().newCommand("test-command").invokeHandler(() -> println("meow").register()
```

**Comment by phase** [comment]
Or an object. Objects are nice. We're using Java for a reason.

**Comment by PizzaCrust** [comment]
How would you propose to use an object for this?

**Comment by phase** [comment]
Not called `Command` but something like

```
public abstract class Command {

    public abstract String getName();

    public abstract void invoke(String[] args, CommandSender sender);

}

```

Which plugins can extend like

```
public class TestCommand extends Command {

    public String getName() { return "test"; }

    public void invoke(String args[], CommandSender sender) { System.out.println(Arrays.toString(args)); }

}

```

File: src/main/java/org/fountainmc/api/command/CommandBuilder.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
+package org.fountainmc.api.command;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class CommandBuilder {
```
Comment (phase) [inline src/main/java/org/fountainmc/api/command/CommandBuilder.java]: A builder to build an interface is redundant when you can create one class that can do the same thing.

**Comment by PizzaCrust** [comment]
What's wrong with using builders?

``` java
Fountain.getServer().getCommandManager()
   .newCommand("test")
   .invokeHandler((args, sender) -> {})
   .generate();
```

**Comment by PizzaCrust** [comment]
Ok, removing builders.

File: src/main/java/org/fountainmc/api/command/RegistryHandler.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+/**
+ * Represents a registry handler.
+ * A registry handler is a listener intended for internal use for implementations.
+ * Allows to register plugins on {@link CommandManager}#registerCommands(Object source) and {@link CommandManager}#registerCommand(ICommand command).
+ */
+public interface RegistryHandler {
+    void onRegister(AbstractCommand iCommand);
+
+    void onRegister(CommandManager.CommandHandler commandHandler);
+}
```
Comment (phase) [inline src/main/java/org/fountainmc/api/command/RegistryHandler.java:14]: You need some more new lines.

```
public class H {

    public void method() {

    }

}

```

Read the contributing file and add the formatters.

**Comment by PizzaCrust** [comment]
Replaced with an abstract class.

**Comment by PizzaCrust** [comment]
Added the lines.

**Comment by phase** [comment]
Merged in https://github.com/FountainMC/FountainAPI/commit/73b702e2e72ceb808a32c8351d3032ffa93a0247

**Pull request closed without merging**